libctru/source/services/iruser.c

```diff
@@ -0,0 +1,569 @@
+
+/// Read and parse the ir:USER service status data from shared memory.
+IRUSER_StatusInfo iruserGetStatusInfo() {
+    void* shared_mem = iruserSharedMem;
+    IRUSER_StatusInfo status_info;
+    // copy over data
+    memcpy(&status_info, shared_mem, sizeof(status_info));
+
+    return status_info;
+}
+
+Result iruserGetCirclePadProState(IRUSER_Packet* packet, circlePadProInputResponse* response) {
+    if (!packet) goto failure;
+    if (!packet->payload) goto failure;
+    if (packet->payload_length != 6) goto failure;
+    if (packet->payload[0] != CIRCLE_PAD_PRO_INPUT_RESPONSE_PACKET_ID) goto failure;
+    response->cstick.csPos.dx = (u16)(packet->payload[1] | ((packet->payload[2] & 0x0F) << 8));
+    response->cstick.csPos.dy = (u16)(((packet->payload[2] & 0xF0) >> 4) | ((packet->payload[3]) << 4));
+    response->status_raw = packet->payload[4];
+    response->unknown_field = packet->payload[5];
+    return 0;
+    failure: 
+    memset(response, 0, sizeof(circlePadProInputResponse));
+    return -1;
+}
+
+Result iruserCirclePadProCStickRead(circlePosition *pos) {
+    // Result ret = 0;
+    u32 n = 0;
+    IRUSER_Packet* packet = iruserGetPackets(&n);
+    if (n == 0) return -1;
+    if (packet[n-1].payload_length != 6) return RS_INVALIDSTATE;
+    if (packet[n-1].payload[0] != CIRCLE_PAD_PRO_INPUT_RESPONSE_PACKET_ID) return RS_INVALIDSTATE;
+    pos->dx = (s16)(packet[n-1].payload[1] | ((packet[n-1].payload[2] & 0x0F) << 8));
+    pos->dy = (s16)(((packet[n-1].payload[2] & 0xF0) >> 4) | ((packet[n-1].payload[3]) << 4));
+
+    free(packet[n-1].payload);
+    free(packet);
+    IRUSER_ReleaseReceivedData(n);
+    return 0;
+}
+
+// since data buffer is a ring buffer, this helper macro makes it easier to access the data
+#ifndef IRUSER_PACKET_DATA
+#define IRUSER_PACKET_DATA(i) (iruserRecvPacketDataBuffer + (inf.offset + i) % iruserRecvPacketDataBufferSize)
+
+static bool iruserParsePacket(size_t index, IRUSER_Packet* packet) {
+    if (packet == NULL) return false;
+    IRUSER_PacketInfo inf = iruserRecvPacketInfoBuffer[index % iruserRecvPacketCount];
+    packet->magic_number = *IRUSER_PACKET_DATA(0);
+    packet->destination_network_id = *IRUSER_PACKET_DATA(1);
+    bool large = *IRUSER_PACKET_DATA(2) & 0x40;
+    u32 payload_offset = 0;
+    if (large) {
+        packet->payload_length = (*IRUSER_PACKET_DATA(2) << 8) + *IRUSER_PACKET_DATA(3);
+        packet->payload = malloc(packet->payload_length);
+        payload_offset = 4;
+    } else {
+        packet->payload_length = *IRUSER_PACKET_DATA(2);
+        packet->payload = malloc(packet->payload_length);
+        payload_offset = 3;
+    }
+    for (size_t i = 0; i < packet->payload_length; i++) {
+        packet->payload[i] = *IRUSER_PACKET_DATA(i + payload_offset);
+    }
+    packet->checksum = *IRUSER_PACKET_DATA(packet->payload_length + payload_offset);
+    // check the checksum
+    u8 checksum = crc8ccitt(IRUSER_PACKET_DATA(0), packet->payload_length + payload_offset);
```

Comment:
bogus, should be u8 checksum = crc8ccitt(packet->payload, packet->payload_length);